### PR TITLE
fix(api-gateway): SQL-runner: permissions fixes

### PR DESCRIPTION
**Pull request by wolfeeeg**
**Check List**
- [ ] Tests has been run in packages where changes made if available
- [ ] Linter has been run for changed code
- [ ] Tests for the changes have been added if not covered yet
- [ ] Docs have been added / updated if required

**Issue Reference this PR resolves**

[For example #12]

**Description of Changes Made (if issue reference is not provided)**

[Description goes here]

### CODE REVIEW

File: packages/cubejs-api-gateway/src/helpers/sqlQueryHandler.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+  if (/^(EXPLAIN)\b/i.test(sql)) {
+    return SQlRunnerQueryType.Explain;
+  }
+
+  return null;
+};
+
+// check query string for scope permissions
+// an example scope with permissions: ['sql-runner-permissions:create,update,delete,select']
```
Comment (wolfeeeg) [inline packages/cubejs-api-gateway/src/helpers/sqlQueryHandler.ts:78]: decided use more complex cope like `sql-runner-permissions:create,update,delete,select`

File: packages/cubejs-api-gateway/src/helpers/sqlQueryHandler.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  const permissions = p.split(',');
+
+  const statementType = getStatementType(query);
+
+  if (statementType !== null && permissions.includes(statementType)) {
+    return true;
+  }
+
+  return false;
+};
```
Comment (MikeNitsenko) [inline packages/cubejs-api-gateway/src/helpers/sqlQueryHandler.ts:103]: Please consider reviewing the code for better readability.
Also, please change the name of the `isQueryHasPermission` to smth like `isQueryAllowed`

File: packages/cubejs-api-gateway/src/helpers/sqlQueryHandler.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  Explain = 'explain',
+}
+
+interface SQLRunnerQueryType {
+  type: SQLQueryType;
+  regex?: RegExp;
+  shouldAddLimit?: boolean;
+}
+
+function createSQLRunnerQueryType({
```
Comment (buntarb) [inline packages/cubejs-api-gateway/src/helpers/sqlQueryHandler.ts:20]: There is no sense in this function. You can just add the dictionary like this:
```
const dict = {
  select: {
    re: new RegExp('^(select)\\b', 'i'),
    requiredLimit: true,
  },
  ...
}
```

File: packages/cubejs-api-gateway/src/helpers/sqlQueryHandler.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  Insert = 'insert',
+  Update = 'update',
+  Delete = 'delete',
+  Explain = 'explain',
+}
+
+interface SQLRunnerQueryType {
+  type: SQLQueryType;
+  regex?: RegExp;
+  shouldAddLimit?: boolean;
```
Comment (buntarb) [inline packages/cubejs-api-gateway/src/helpers/sqlQueryHandler.ts:17]: Please, rename it to `requiredLimit`.

File: packages/cubejs-api-gateway/src/helpers/sqlQueryHandler.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+export function getSQLRunnerQueryType(sql: string): SQLRunnerQueryType | null {
+  const sqlRunnerQueryTypes: SQLRunnerQueryType[] = [];
+
+  for (const type of Object.values(SQLQueryType)) {
+    const option = createSQLRunnerQueryType({ type });
+    sqlRunnerQueryTypes.push(option);
+  }
+
+  for (const queryType of sqlRunnerQueryTypes) {
```
Comment (buntarb) [inline packages/cubejs-api-gateway/src/helpers/sqlQueryHandler.ts:45]: In a case of using dictionary you can simplify this.

File: packages/cubejs-api-gateway/src/types/request.ts [hunk 6]
```diff
@@ -19,9 +19,10 @@ interface RequestContext {
   securityContext: any;
   requestId: string;
   signedWithPlaygroundAuthSecret?: boolean;
-  appName?: string,
-  protocol?: string,
-  apiType?: string,
+  appName?: string;
+  protocol?: string;
+  apiType?: string;
+  sqlRunnerShouldAddLimit?: boolean;
```
Comment (buntarb) [inline packages/cubejs-api-gateway/src/types/request.ts:25]: Please, rename it to `requiredLimit`.

File: packages/cubejs-api-gateway/src/helpers/sqlQueryHandler.ts [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  return {
+    type,
+    regex,
+    shouldAddLimit,
+  };
+}
+
+export function getSQLRunnerQueryType(sql: string): SQLRunnerQueryType | null {
```
Comment (buntarb) [inline packages/cubejs-api-gateway/src/helpers/sqlQueryHandler.ts:37]: Actually, you can rename it to `isRequiredLimit` and return boolean value from it.

**Comment by wolfeeeg** [comment]
moved to cloud

**Pull request closed without merging**